R/sim_linelist.R

```diff
@@ -0,0 +1,217 @@
+#'   hosp_rate = age_dep_hosp_rate
+#' )
+sim_linelist <- function(R,
+                         serial_interval,
+                         onset_to_hosp,
+                         onset_to_death,
+                         hosp_rate = 0.2,
+                         hosp_death_rate = 0.5,
+                         non_hosp_death_rate = 0.05,
+                         outbreak_start_date = as.Date("2023-01-01"),
+                         add_names = TRUE,
+                         add_ct = FALSE,
+                         min_chain_size = 10,
+                         population_age = c(1, 90),
+                         case_type_probs = c(
+                           suspected = 0.2,
+                           probable = 0.3,
+                           confirmed = 0.5
+                         ),
+                         config = create_config(),
+                         ...) {
+  chkDots(...)
+
+  serial_interval <- as.function(serial_interval, func_type = "generate")
+  onset_to_hosp <- as.function(onset_to_hosp, func_type = "generate")
+  onset_to_death <- as.function(onset_to_death, func_type = "generate")
+
+  .check_sim_input(
+    sim_type = "linelist",
+    R = R,
+    serial_interval = serial_interval,
+    outbreak_start_date = outbreak_start_date,
+    min_chain_size = min_chain_size,
+    onset_to_hosp = onset_to_hosp,
+    onset_to_death = onset_to_death,
+    add_names = add_names,
+    add_ct = add_ct,
+    case_type_probs = case_type_probs,
+    hosp_rate = hosp_rate,
+    hosp_death_rate = hosp_death_rate,
+    non_hosp_death_rate = non_hosp_death_rate,
+    population_age = population_age
+  )
+
+  if (is.data.frame(population_age)) {
+    population_age <- .check_age_df(population_age)
+    age_range <- c(
+      min(population_age[, "min_age"]),
+      max(population_age[, "max_age"])
+    )
+  } else {
+    age_range <- population_age
+  }
+  if (is.data.frame(hosp_rate)) {
+    hosp_rate <- .check_rate_df(
+      hosp_rate,
+      age_range = age_range
+    )
+  }
+  if (is.data.frame(hosp_death_rate)) {
+    hosp_death_rate <- .check_rate_df(
+      hosp_death_rate,
+      age_range = age_range
+    )
+  }
+  if (is.data.frame(non_hosp_death_rate)) {
+    non_hosp_death_rate <- .check_rate_df(
+      non_hosp_death_rate,
+      age_range = age_range
+    )
+  }
+
+  chain <- .sim_bp_linelist(
+    R = R,
+    serial_interval = serial_interval,
+    outbreak_start_date = outbreak_start_date,
+    min_chain_size = min_chain_size,
+    population_age = population_age,
+    config = config
+  )
+
+  linelist <- .sim_clinical_linelist(
+    chain = chain,
+    onset_to_hosp = onset_to_hosp,
+    onset_to_death = onset_to_death,
+    hosp_rate = hosp_rate,
+    hosp_death_rate = hosp_death_rate,
+    non_hosp_death_rate = non_hosp_death_rate,
+    outbreak_start_date = outbreak_start_date,
+    add_names = add_names,
+    add_ct = add_ct,
+    case_type_probs = case_type_probs,
+    config = config
+  )
```

Comment:
Upon reading more of the code, I think I understand the rationale behind this separation: `.sim_bp_linelist()` create columns relevant for contacts and linelist while `.sim_clinical_linelist()` focuses on columns only relevant for the linelist.
I suspect these wrappers were created to reduce duplication across the `sim_*()` functions.

However, I don't think the goal is fully achieved as there is still a lot of duplication.

I would like to consider a different refactoring approach:

All `sim_*()` functions call a `sim_internal()` function:

```r
sim_internal <- function(out = c("linelist", "contacts")) {

  out <- match.arg(out, several.ok = TRUE)

  # arg checking

  # simulate branching process
  # ...

  if ("linelist" %in% out) {
    # simulate linelist specific columns
  }
  if ("contacts" %in% out) {
    # simulate contact specific columns
  }

 # return df or list depending on out value
}
```

This doesn't result in more nesting since we can then remove the `sim_utils` and `sim_contact_tbl()` functions and call directly the `add_col` function.

I find it clearer because you don't have to guess where a specific column will be computed ("is it `.sim_bp_linelist()` or  `.sim_clinical_linelist()`?"). Everything happens clearly in the `sim_internal()` function and the exported `sim_*()` functions are just very thin wrappers around it. 

If performance is not an issue, this `sim_internal()` could also be `sim_outbreaks()`. `sim_linelist()` and `sim_contacts()` would then just keep only the relevant df and drop the other.

Reply:
I don't expect any of this refactoring to result in user-facing changes so it can wait until after the first release.

Reply:
Logged in issue #42. As you suggest, I will tackle this in a future version.